### PR TITLE
ARIA-140 Version utils

### PR DESCRIPTION
**Pull request by tliron**

### CODE REVIEW

File: aria/utils/versions.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+                return _NULL
+            qualifier = value
+    else:
+        # Version strings with no qualifiers are higher
+        qualifier = _INF
+
+    return primary, qualifier
+
+
+VersionString.NULL = VersionString()
```
Comment (AviaE) [inline aria/utils/versions.py]: What is the purpose of this NULL? I don't see any use of it except in the tests.

File: aria/modeling/service_template.py [hunk 2]
```diff
@@ -33,7 +33,8 @@
 from ..parser import validation
```
Comment (AviaE) [inline aria/modeling/service_template.py]: Please expand the commit message by adding information on the purpose of your change/addition.

File: aria/utils/versions.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+"""
+General-purpose version string handling
+"""
+
+import re
+
+
+_INF = float('inf')
+
+_NULL = (), _INF
```
Comment (tliron) [inline aria/utils/versions.py]: Because "Numeric qualifiers will always be greater than prefixed integer qualifiers, e.g. "1.1-1" > "1.1-beta1". The qualifier is a float that is an extra, but if the qualifier is not present "infinity" should always win over it. It shouldn't matter what's "more natural" seeming: this is an underscored internal var that's never seen by users of the lib.

File: aria/utils/versions.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+def parse_version_string(version): # pylint: disable=too-many-branches
+    """
+    Parses a version string.
+
+    :param version: The version string
+    :returns: The primary tuple and qualifier float
+    :rtype: ((int), float)
+    """
+
+    if version is None:
```
Comment (tliron) [inline aria/utils/versions.py]: Already in line #64 do you see the default to be None. But anyway, this is a user-exposed function and None values do occur in the wild.

File: aria/utils/versions.py (hunk 3 above)
Comment (tliron) [inline aria/utils/versions.py]: Because "Numeric qualifiers will always be greater than prefixed integer qualifiers, e.g. "1.1-1" > "1.1-beta1". The qualifier is a float that is an extra, but if the qualifier is not present "infinity" should always win over it. It shouldn't matter what's "more natural" seeming: this is an underscored internal var that's never seen by users of the lib. Users instead use `VersionString.NULL` or can init `VersionString` with None.

File: aria/utils/versions.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        return _NULL
+    version = unicode(version)
```
Comment (tliron) [inline aria/utils/versions.py]: It is true, there is no test for this function specifically with a None value. I will add it.

File: aria/utils/versions.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+    # Split to primary and qualifier on '-'
+    split = version.split('-', 2)
+    if len(split) == 2:
+        primary, qualifier = split
+    else:
```
Comment (tliron) [inline aria/utils/versions.py]: Oops, yes, changed to rsplit on 1.

File: aria/utils/versions.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        qualifier = None
+
+    # Parse primary
+    split = primary.split('.')
+    primary = []
+    for element in split:
+        if _DIGITS_RE.match(element) is None:
+            # Invalid version string
+            return _NULL
+        try:
```
Comment (tliron) [inline aria/utils/versions.py]: Maybe it's too many digits for the current Python platform's size of int? Seems to me that it's best not to assume it would always work. Utility methods need to be especially robust.

File: aria/utils/versions.py (hunk 1 above)
Comment (tliron) [inline aria/utils/versions.py]: It's a constant for the user to use in any situation where they need a `NULL` value. It's customary to provide singletons for efficiency.

File: aria/modeling/service_template.py (hunk 2 above)
Comment (tliron) [inline aria/modeling/service_template.py]: I don't understand. What does this have to do with this line in the code? Where in the commit message? We have limited our commit messages to one line with reference to the JIRA for more information. I know you tend to add more info, but the rest of us don't normally.

File: aria/modeling/service_template.py [hunk 8]
```diff
@@ -33,7 +33,8 @@
...
 from ..parser.consumption import ConsumptionContext
 from ..parser.reading import deepcopy_with_locators
-from ..utils import collections, formatting, console
+from ..utils import (collections, formatting, console)
```
Comment (tliron) [inline aria/modeling/service_template.py]: We have discussed this and have agreed that it can be in one line if it fits. We have a lot of code that is like this.

File: aria/modeling/service_template.py [hunk 9]
```diff
@@ -2131,13 +2132,15 @@ def resolve(self, model_storage):
...
         plugins = model_storage.plugin.list()
         matching_plugins = []
-        for plugin in plugins:
-            # TODO: we need to use a version comparator
-            if (plugin.name == self.name) and \
-                    ((self.version is None) or (plugin.package_version >= self.version)):
-                matching_plugins.append(plugin)
+        if plugins:
+            for plugin in plugins:
+                if (plugin.name == self.name) and \
```
Comment (tliron) [inline aria/modeling/service_template.py]: I strongly disagree, especially in this case when there is `or` and `and` in the same clause. Do you otherwise expect the reader to know the priority of logical binary operations? Parentheses remove any potential confusion.

**Comment by tliron** [comment]
Responses.

File: aria/utils/versions.py (hunk 4 above)
Comment (tliron) [inline aria/utils/versions.py]: This is a user-exposed function and None values do occur in the wild.

File: aria/utils/versions.py (hunk 5 above)
Comment (tliron) [inline aria/utils/versions.py]: It is true, there is no tests for this function specifically with a None value. I will add it.

File: aria/utils/versions.py (hunk 3 above)
Comment (tliron) [inline aria/utils/versions.py]: Because "Numeric qualifiers will always be greater than prefixed integer qualifiers, e.g. "1.1-1" > "1.1-beta1". The qualifier is a float that is an extra, but if the qualifier is not present "infinity" should always win over it. It shouldn't matter what's "more natural" seeming: this is an underscored internal var that's never seen by users of the lib. Users instead use `VersionString.NULL` or can init `VersionString` with None.

**Comment by AviaE** [comment]
@tliron and I discussed his replies via zoom, and this pull request is free to merge (with minor changes).
